### PR TITLE
[Custom Fields] Add custom field forms

### PR DESCRIPTION
**Pull request by markkaylor**
### What does it do?

- Adds the user registered forms on custom fields
- Handles save with customField attribute
- Handles validation on the forms
- Adds a new sections to the advanced tab form for other plugins (i.e. i18n) to add custom inputs

### How to test it?
- Run `exmaples/getstarted` with `yarn develop --watch-admin`
- Go to any contentType and click add attribute
- Select the custom tab and click the "Color" custom field, you should see the forms as they are registered in `examples/getstarted/src/plugins/mycustomfields/admin/src/index.js`
- Don't fill the form and hit finish, you should get a validation error on both inputs. The error message on the custom field should be what's registered in examples/getstarted
- Name the field something that already exists on the content type, you should get a validation error
-  Fill out the form and click finish, it should add the attribute to the list
- Click save, it should save to the schema.json:
- 
```
[name]: {
  type: "customField"
  customField: "plugin::mycustomfields.color"
  options: {
     format: 'rgb' | 'hex'
  }
}
```

- Go to any content type where i18n is enabled
- Click the create another field button
- Select the custom tab, choose color, and go to the advanced tab. You should see the Extended settings section with a - checkbox for "Enable localization for this field"
- It should be enabled by default if the content-type has i18n turned on
- Saving a custom field attribute with i18n should save a schem.json like:
```
[name]: {
  "pluginOptions": {
    "i18n": {
      "localized": true | false
    }
  },
  "type": "customField",
  "options": {
     "format": "hex" | "rbga"
  },
  "customField": "plugin::mycustomfields.color"
}
```

### CODE REVIEW

**Comment by markkaylor** [comment]
This is still a WIP but I think I could use some feedback already @gu-stav (@soupette if you have the time 🙏 ). 

Based on notes from Soupette, I tried to add instead of modify:
- Created a new `modalType` for customFields
- Created new actions for customFields in the DataManager: `ON_ADD_CUSTOM_FIELD_ATTRIBUTE` and `ON_EDIT_CUSTOM_FIELD_ATTRIBUTE`. Here I am not confident about the shared logic between these actions and `ADD_ATTRIBUTE` AND `EDIT_ATTRIBUTE`

File: packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js
Comment (markkaylor) [inline packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:21]: @soupette is this what you had in mind?

**Comment by soupette** [comment]
> This is still a WIP but I think I could use some feedback already @gu-stav (@soupette if you have the time 🙏 ).
> 
> Based on notes from a meeting with Soupette (shared on notion), I tried to add instead of modify:
> 
> * Created a new `modalType` for customFields
> * Created new actions for customFields in the DataManager: `ON_ADD_CUSTOM_FIELD_ATTRIBUTE` and `ON_EDIT_CUSTOM_FIELD_ATTRIBUTE`. Here I am not confident about how I am sharing logic between these actions and `ADD_ATTRIBUTE` and `EDIT_ATTRIBUTE` (mainly, do I have everything I need for custom fields...seems to be working 🤔 )

I'll take a look at it next week!

**Comment by markkaylor** [comment]
Awesome thanks for your help @soupette !

File: packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js
Comment (soupette) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:21]: I haven't reviewed the code yet but I think you missed the validation part, as you can create a custom field called slug for address which already exists.

Do you want me to take a look at the code first without testing the behavior?

Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:21]: Yeah I haven't added the validation yet. I worked on it yesterday and will probably open a new PR for that today. I'll also need to make sure plugins can inject inputs into the forms. If you prefer to wait until I've added all, no problem I can ping you once ready. But if you are able to give this a once over as is that could also be helpful. Whatever works best for you.

Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:21]: Here's the validation: https://github.com/strapi/strapi/pull/13935

Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:21]: Here's the injected inputs: https://github.com/strapi/strapi/pull/13940

File: packages/core/content-type-builder/admin/src/components/FormModal/forms/utils/addItemsToFormSection.js
Comment (soupette) [inline packages/core/content-type-builder/admin/src/components/FormModal/forms/utils/addItemsToFormSection.js:1]: Can you add a test?

File: packages/core/content-type-builder/admin/src/components/FormModal/index.js
Comment (soupette) [inline packages/core/content-type-builder/admin/src/components/FormModal/index.js:300]: 👍

**Review by soupette — COMMENTED** [review]
So far this is a really good work I just have a few concerns expressed in my comments.

File: packages/core/content-type-builder/admin/src/components/DataManagerProvider/reducer.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/DataManagerProvider/reducer.js:95]: Not sure I understand, I'm not passing it I'm destructuring the returned value from  `getAttributeUpdate`.

File: packages/core/content-type-builder/admin/src/components/DataManagerProvider/reducer.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/DataManagerProvider/reducer.js:240]: I'm still using `forTarget` but destructuring it inside `getEditAttributeUpdate` instead of here

File: packages/core/content-type-builder/admin/src/components/DataManagerProvider/reducer.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/DataManagerProvider/reducer.js:291]: gotcha ok, I was try to respect DRY but if it's ok to repeat the code I could just do it that way then.

File: packages/core/content-type-builder/admin/src/components/DataManagerProvider/index.js
Comment (gu-stav) [reply packages/core/content-type-builder/admin/src/components/DataManagerProvider/index.js:176]: While I agree the method is a good idea, I'd prefer if you'd change the signature from args to a single object, because it contains so many params. It would make it easier in the context of where it is used to understand what each param means. Wdyt?

File: packages/core/content-type-builder/admin/src/components/FormModal/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/FormModal/index.js:559]: Now I see the usage, I'd update my comment from above the recommend to split these up. It would also simplify the signature:

```
        if (actionType === 'edit') {
          editCustomFieldAttribute({ });
        } else {
           addCustomFieldAttribute({ });
        }
        
```

File: examples/getstarted/src/plugins/mycustomfields/admin/src/index.js
Comment (gu-stav) [inline examples/getstarted/src/plugins/mycustomfields/admin/src/index.js:120]: I've mentioned it on Slack already and don't want to bikeshed here, so please feel free to ignore if you think it is not a good idea:

1. I'd rename this method to `validate()`, to make it more imperative
2. I'd change the signature to return a `boolean` instead, because it might feel more intuitive to our users. They'd still have the chance to validate the `yup` schema.

File: packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js:388]: Woow, this component is hell 🔥 . Do you think this is a good chance to split it up a bit? 👀

File: packages/core/content-type-builder/admin/src/components/FormModalHeader/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/FormModalHeader/index.js:114]: Mh, I'm not sure I like that. Why? Because you already introduce some custom fields conditionals here, so why not picking the icon here from `customFieldsRegistry` and then pass it down as `type={}`?

File: packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:20]: I think this should be called `isCustomField` or something like that.

File: packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js
Comment (gu-stav) [inline packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:45]: Maybe I am missing something here, but what do you need `type`, `name` and `step` for?

Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:45]: They are needed depending on the returned value from `getModalTitleSubheader`. For customFields we need `type` and `name`.`type` is used when the `actionType` is create and `name` is used when the `actionType` is edit

File: packages/core/content-type-builder/admin/src/components/FormModalHeader/index.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalHeader/index.js:114]: I think `type` expects a string that it uses as a lookup in `AttributeIcon`. Since an icon isn't required on a custom field we want to use that type's icon if the user did not register their custom field with an icon. That logic is handled in the `AttributeIcon` component

File: packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:20]: I think it should remain `type` since it doesn't return a boolean and the object shorthand can be used in the formatMessage variables object. Maybe you meant just below for `modalType === 'customField'`?

Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:20]: Or maybe something like this 🤷‍♂️ ?
```javascript
const isCustomField = modalType === 'customField';
const intlLabel = isCustomField
  ? customField.intlLabel
  : { id: getTrad(`attribute.${attributeType}`)};
//...
{
  type: upperFirst(formatMessage(intlLabel)),
  name: upperFirst(attributeName),
  step,
}
```

Comment (gu-stav) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js]: I don't see the original code any longer, but when I reviewed it I had the impression that `type` is a boolean holding the information whether it is a custom field or not. If that is not the case, feel free to resolve this discussion :)

You could also shorten this to:

```js
{
  type: upperFirst(formatMessage(customField?.intlLabel ?? { id: getTrad(`attribute.${attributeType}`)})),
  name: upperFirst(attributeName),
  step,
}
```

**Comment by gu-stav** [comment]
@markkaylor I think I closed all threads that have been resolved/ discussed. Let me know once you need another review (and the approval :)).

File: packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js:388]: I don't mind cleaning this component up a bit, but what did you have in mind, moving each button group to its own component?

File: examples/getstarted/src/plugins/mycustomfields/admin/src/index.js
Comment (markkaylor) [reply examples/getstarted/src/plugins/mycustomfields/admin/src/index.js:120]: I think your point is valid, but after looking a bit more at this and discussing with Rémi I don’t think we will make this change for a few reasons:

1. This design was detailed in the RFC which was left in review for over a month
2. We couldn’t return just true or false, we would need the name of the input and an error message as well. Throwing a general error message is not a great UX.
3. Forcing a user to use Yup is not necessarily a bad thing. It is consistent with how validation is used in Strapi and it is not a difficult API to use.

For the naming you also make a good argument but we will also stick with `validator` since we are using an existing API, the content-type builder’s forms API, that has already established this convention.

File: packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js
Comment (gu-stav) [reply packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js:388]: Yes, I think that would be an improvement in terms of readability. Wdyt? Could also be a task after this PR to keep the scope a bit smaller ...

Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js:388]: In terms of timing it might be better to leave this as is until the alpha and make the improvement before the beta

File: packages/core/content-type-builder/admin/src/components/FormModal/forms/utils/addItemsToFormSection.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/FormModal/forms/utils/addItemsToFormSection.js:18]: I think you mean "when a sectionTitle is present" for this comment

File: packages/core/content-type-builder/admin/src/components/FormModal/forms/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/FormModal/forms/index.js:12]: We can maybe move this to the utils folder since it already exists and this file is super long already

File: packages/core/content-type-builder/admin/src/components/FormModal/forms/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/FormModal/forms/index.js:70]: Where is that logic handled for non-custom field attributes?

File: packages/core/content-type-builder/admin/src/components/FormModal/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/FormModal/index.js:110]: Doesn't respect alphabetical order

File: packages/core/content-type-builder/admin/src/components/FormModal/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/FormModal/index.js:1058]: Typo: cusom instead of custom

File: packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/FormModalEndActions/index.js:41]: custom -> custom

**Review by remidej — CHANGES_REQUESTED** [review]
That's a big one, nice work! One thing I notice but I'm not sure if it's normal: we can't add a custom field in a component directly (in the CTB outside the context of a single or collection type).

File: packages/core/content-type-builder/admin/src/components/FormModal/forms/index.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModal/forms/index.js:70]: I believe it is this: https://github.com/strapi/strapi/blob/aaf64cefa0365ccc900a89a978b502a8ccb77ba6/packages/core/content-type-builder/admin/src/components/FormModal/forms/index.js#L132-L148

**Comment by markkaylor** [comment]
@kibblerz we are aiming to release an alpha version by the end of August

File: packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js
Comment (remidej) [inline packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:36]: In theory I don't think this would be needed since we [throw](https://github.com/strapi/strapi/blob/features/custom-fields/packages/core/admin/admin/src/core/apis/CustomFields.js#L21) when intlLabel is not provided

**Review by remidej — CHANGES_REQUESTED** [review]
Thanks for making the changes!

I still have 1 bug: when I try to add a map custom field, `customField.options is undefined`

My other comment is optional, do what you think is best

File: packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js
Comment (markkaylor) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js:36]: We need to return the attribute translation when the modal is not for a custom field. So I think the ternary I had before is probably more readable.
 
```
const intlLabel = modalType === 'customField' ? customField.intlLabel : { id: getTrad(`attribute.${attributeType}`) }
```

**Comment by markkaylor** [comment]
> I still have 1 bug: when I try to add a map custom field, `customField.options is undefined`

Seen that one before, thought I had fixed it... Should be good now I think.

**Review by remidej — APPROVED** [review]
🚀

File: packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js
Comment (remidej) [reply packages/core/content-type-builder/admin/src/components/FormModalSubHeader/index.js]: I misunderstood what the condition was for sorry

**Pull request merged**
